### ISSUE TITLE
`ShapeMismatchBinaryOp` occurred in the convolutional network while executing `backward_step`

### ISSUE THREAD

**Issue by antodyn** [huggingface/candle#1212]
Why is my convolutional network not backpropagating properly? Is there a problem with my model building code? The model code can be checked by the compiler, but when training, an error occurs when loss.backward_step is executed. Here is the simplified code.
```
use candle_core::{DType, Device, Result, Tensor, D};
use candle_nn::{
    BatchNorm, BatchNormConfig, Conv2d, Conv2dConfig, Linear, Module, Optimizer, VarBuilder,
    VarMap, SGD,
};

#[derive(Debug)]
pub struct ModelTest {
    conv1: Conv2d,
    conv2: Conv2d,
    bn: BatchNorm,
    head: Linear,
}
impl ModelTest {
    pub fn new(vb: candle_nn::VarBuilder, out_channels: usize) -> candle_core::Result<Self> {
        let conv1_cfg = Conv2dConfig {
            stride: 2,
            padding: 3,
            ..Default::default()
        };
        let conv1 = candle_nn::conv2d_no_bias(3, out_channels, 7, conv1_cfg, vb.pp("conv2d1"))?;

        let conv2_cfg = Conv2dConfig {
            stride: 2,
            padding: 1,
            ..Default::default()
        };
        let conv2 =
            candle_nn::conv2d_no_bias(out_channels, out_channels, 3, conv2_cfg, vb.pp("conv2d2"))?;

        let bn_cfg = BatchNormConfig {
            eps: 1e-5,
            affine: true,
            remove_mean: false,
        };
        let bn = candle_nn::batch_norm(out_channels, bn_cfg, vb.pp("bn"))?;
        let head = candle_nn::linear(20 * 64 * 64, 4, vb.pp("head"))?;

        Ok(Self {
            conv1,
            conv2,
            bn,
            head,
        })
    }
}
impl Module for ModelTest {
    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
        let xs = self.conv1.forward(xs).unwrap();
        let xs = self.bn.forward(&xs).unwrap();
        let xs = self.conv2.forward(&xs).unwrap();
        let xs = self.bn.forward(&xs).unwrap();
        let xs = xs.reshape((1, 20 * 64 * 64)).unwrap();
        self.head.forward(&xs)
    }
}

fn main() -> Result<()> {
    let device = Device::cuda_if_available(0)?;
    let mut varmap = VarMap::new();
    let vb = VarBuilder::from_varmap(&mut varmap, DType::F32, &device);

    let inputs =
        Tensor::arange(0f32, (3 * 256 * 256) as f32, &device)?.reshape((1, 3, 256, 256))?;
    let labels = Tensor::from_slice(&[0u32], (1,), &device).unwrap();

    let model = ModelTest::new(vb, 20)?;
    let mut opt = SGD::new(varmap.all_vars(), 0.05).unwrap();

    let logits = model.forward(&inputs).unwrap();
    let log_sm = candle_nn::ops::log_softmax(&logits, D::Minus1).unwrap();
    let loss = candle_nn::loss::nll(&log_sm, &labels).unwrap();
    opt.backward_step(&loss).unwrap(); // Error in here

    Ok(())
}
```
The result of execution is:
```
thread 'main' panicked at src/main.rs:73:30:
called `Result::unwrap()` on an `Err` value: ShapeMismatchBinaryOp { lhs: [20, 20, 3, 3], rhs: [20, 20, 4, 4], op: "add" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I found that the error was triggered in [`candle-core/src/backprop.rs`](https://github.com/huggingface/candle/blob/498c50348ce13456d683c987ad9aef319a45eb4a/candle-core/src/backprop.rs#L241C1-L241C65). 

I'm new to candle and I'm not sure if I've figured out the cause of the problem, but right now I don't know how to update my code to keep going. I look forward to your reply and send my most sincere thanks.

**Comment by LaurentMazare** [comment]
Thanks for reporting this and for providing a simple repro. Turns out that it's a bug on the candle side which happens when using strides different from 1 but only on certain shapes (we already had a test for the conv2d gradient with stride but it didn't trigger for this).
Hopefully this should be fixed in #1214 , this PR also includes a test that was failing in the same way as yours and checks that the resulting gradients are in line with what PyTorch returns. Please give this a spin and let us know how it goes.

**Comment by antodyn** [comment]
Thank you very much, it works very well on my code.

> Thanks for reporting this and for providing a simple repro. Turns out that it's a bug on the candle side which happens when using strides different from 1 but only on certain shapes (we already had a test for the conv2d gradient with stride but it didn't trigger for this). Hopefully this should be fixed in #1214 , this PR also includes a test that was failing in the same way as yours and checks that the resulting gradients are in line with what PyTorch returns. Please give this a spin and let us know how it goes.

**Comment by LaurentMazare** [comment]
Great, closing this for now but feel free to re-open / open a new issue if needed.
